File: src/api/agents.ts
import { generatedGet } from './generated/client';

export type AgentsQuery = {
  selected_month: string;
  firma?: string;
  regional?: string;
  asm?: string;
  site_code?: string;
  agent?: string;
  search?: string;
};

export interface AgentsOverviewResponse {
  // Snapshot
  active_count: number;
  new_count: number;
  reactivated_count: number;
  left_this_month_count: number;
  retention_rate: number | null;
  
  // Health
  total_unique_agents: number;
  avg_seniority_months: number | null;
  stability_rate: number | null;
  churned_total_count: number;
}

export interface StoreCoverageItem {
  site_code: string;
  locatie: string;
  firma: string;
  regional: string;
  asm: string;
  status: 'covered' | 'uncovered' | 'closed' | 'inactive';
  agent_count: number;
  has_changes: boolean;
  previous_agent_count: number;
  added_agents_count: number;
  removed_agents_count: number;
  change_reason: string | null;
}

export interface StoreCoverageResponse {
  active_stores_count: number;
  uncovered_stores_count: number;
  closed_stores_count: number;
  modified_stores_count: number;
  items: StoreCoverageItem[];
}

export interface AgentMovementPoint {
  month: string;
  active: number;
  new: number;
  reactivated: number;
  churned: number;
  net_growth: number;
  is_baseline: boolean;
}

export interface AgentMovementResponse {
  history: AgentMovementPoint[];
}

export interface AgentListItem {
  agent: string;
  store_name: string | null;
  firma: string | null;
  active_in_month: boolean;
  is_new: boolean;
  is_reactivated: boolean;
  total_sales: number;
  total_quantity: number;
  current_status: string;
}

export interface AgentListResponse {
  items: AgentListItem[];
}

export interface AgentProfileResponse {
  agent: string;
  first_seen_month: string;
  last_seen_month: string;
  active_months_count: number;
  distinct_store_count: number;
  distinct_firma_count: number;
  distinct_regional_count: number;
  distinct_asm_count: number;
  months_since_last_seen: number;
  reactivation_count: number;
  longest_active_streak: number;
  career_total_sales: number;
  career_total_quantity: number;
  avg_monthly_sales: number;
  best_month: string | null;
  best_month_sales: number;
  current_status: string;
}

export interface AgentHistoryPoint {
  month: string;
  total_sales: number;
  total_quantity: number;
  receipt_count: number;
  active_store_count: number;
  is_active: boolean;
}

export interface AgentHistoryResponse {
  history: AgentHistoryPoint[];
}

export interface AgentEvaluationOption {
  value: string;
  label: string;
}

export interface AgentEvaluationRow {
  month: string;
  firma: string;
  site_code: string;
  locatie: string;
  regional: string;
  asm: string;
  agent: string;
  total_sales: number;
  total_quantity: number;
  working_days: number;
  store_target: number;
  store_working_days: number;
  target_value: number;
  target_pct: number | null;
  daily_average: number | null;
  peer_daily_average: number | null;
  value_reper: number | null;
  receipt_count: number;
  receipt_2plus_count: number;
  bonuri_pct: number | null;
  focus_quantity: number;
  focus_pct: number | null;
  glass_qty: number;
  premium_glass_qty: number;
  premium_glass_pct: number | null;
  target_points: number;
  daily_points: number;
  value_reper_points: number;
  bonuri_points: number;
  focus_points: number;
  premium_glass_points: number;
  total_points: number;
  has_red_segment: boolean;
  qualifier: string;
}

export interface AgentEvaluationResponse {
  months: AgentEvaluationOption[];
  firmas: AgentEvaluationOption[];
  asms: AgentEvaluationOption[];
  stores: AgentEvaluationOption[];
  rows: AgentEvaluationRow[];
}

export interface AgentEvaluationV2Row {
  month: string;
  firma: string;
  site_code: string;
  locatie: string;
  regional: string;
  asm: string;
  agent: string;
  total_sales: number;
  forecast_sales: number;
  total_quantity: number;
  working_days: number;
  receipt_count: number;
  target_value: number;
  target_source: string;
  target_pct: number | null;
  target_forecast_pct: number | null;
  is_partial: boolean;
  period_month_count: number;
  partial_month_count: number;
  final_month_count: number;
  forecast_factor: number;
  daily_average: number | null;
  daily_reference: number | null;
  daily_reference_type: string;
  daily_vs_reference_pct: number | null;
  value_reper: number | null;
  receipt_2plus_count: number;
  bonuri_pct: number | null;
  focus_quantity: number;
  focus_pct: number | null;
  glass_qty: number;
  premium_glass_qty: number;
  premium_glass_pct: number | null;
  trend_daily_pct: number | null;
  trend_direction: 'up' | 'down' | 'flat';
  eligibility_status: 'eligibil' | 'insuficient';
  confidence_flags: string[];
  target_score: number | null;
  daily_score: number | null;
  bonuri_score: number | null;
  focus_score: number | null;
  premium_glass_score: number | null;
  value_reper_score: number | null;
  total_score: number | null;
  max_score: number;
  rating: string;
}

export interface AgentEvaluationV2Response {
  months: AgentEvaluationOption[];
  firmas: AgentEvaluationOption[];
  asms: AgentEvaluationOption[];
  stores: AgentEvaluationOption[];
  rows: AgentEvaluationV2Row[];
}

export async function fetchAgentsOverview(query: AgentsQuery, signal?: AbortSignal): Promise<AgentsOverviewResponse> {
  return generatedGet('get_agents_overview_api_agents_overview_get', { params: query, signal }) as unknown as AgentsOverviewResponse;
}

export async function fetchAgentsMovement(query: AgentsQuery, signal?: AbortSignal): Promise<AgentMovementResponse> {
  return generatedGet('get_agents_movement_api_agents_movement_get', { params: query, signal }) as unknown as AgentMovementResponse;
}

export async function fetchAgentsList(query: AgentsQuery, signal?: AbortSignal): Promise<AgentListResponse> {
  return generatedGet('get_agents_list_api_agents_list_get', { params: query, signal }) as unknown as AgentListResponse;
}

export async function fetchAgentProfile(agent: string, selectedMonth: string, signal?: AbortSignal): Promise<AgentProfileResponse> {
  return generatedGet('get_agent_profile_api_agents_profile_get', {
    params: { agent, selected_month: selectedMonth },
    signal,
  }) as unknown as AgentProfileResponse;
}

export async function fetchAgentHistory(agent: string, signal?: AbortSignal): Promise<AgentHistoryResponse> {
  return generatedGet('get_agent_history_api_agents_history_get', {
    params: { agent },
    signal,
  }) as unknown as AgentHistoryResponse;
}

export async function fetchStoreCoverage(query: Partial<AgentsQuery>, signal?: AbortSignal): Promise<StoreCoverageResponse> {
  return generatedGet('get_stores_coverage_api_agents_stores_coverage_get', { params: query, signal }) as unknown as StoreCoverageResponse;
}

export async function fetchAgentEvaluation(params: {
  month?: string;
  months?: string;
  firma?: string;
  asm?: string;
  site_code?: string;
} = {}): Promise<AgentEvaluationResponse> {
  return generatedGet('get_agent_evaluation_api_agents_evaluation_get', { params }) as unknown as AgentEvaluationResponse;
}

export async function fetchAgentEvaluationV2(params: {
  month?: string;
  months?: string;
  firma?: string;
  asm?: string;
  site_code?: string;
} = {}): Promise<AgentEvaluationV2Response> {
  return generatedGet('get_agent_evaluation_v2_api_agents_evaluation_v2_get', { params }) as unknown as AgentEvaluationV2Response;
}

File: backend/schemas/agents.py
"""Public API contracts for agent lifecycle and evaluation."""
from __future__ import annotations

from decimal import Decimal
from typing import Annotated, Literal

from pydantic import Field
from schemas.common import StrictApiModel, MonthStr



StoreCoverageStatus = Literal["covered", "uncovered", "closed", "inactive"]
AgentCurrentStatus = Literal["active", "inactive_recent", "churned"]
AgentQualifier = Literal["Excelent", "Foarte Bun", "Bun", "Mediu", "Scazut"]
TargetSource = Literal["partial_agent_target", "allocated_store_target"]
DailyReferenceType = Literal["colegi", "istoric_locatie", "media_manager", "none"]
TrendDirection = Literal["up", "down", "flat"]
EligibilityStatus = Literal["eligibil", "insuficient"]
AgentRating = Literal[
    "Insuficient",
    "Fara scor",
    "Excelent",
    "Foarte Bun",
    "Bun",
    "Risc",
    "Critic",
]
AgentEvaluationPeriod = Annotated[
    str,
    Field(pattern=r"^(?:\d{4}-(?:0[1-9]|1[0-2])(?:\.\.curent)?|custom)$"),
]


class AgentsOverviewResponse(StrictApiModel):
    active_count: int
    new_count: int
    reactivated_count: int
    left_this_month_count: int
    retention_rate: Decimal | None
    total_unique_agents: int
    avg_seniority_months: Decimal | None
    stability_rate: Decimal | None
    churned_total_count: int


class StoreCoverageItem(StrictApiModel):
    site_code: str
    locatie: str
    firma: str
    regional: str
    asm: str
    status: StoreCoverageStatus
    agent_count: int
    has_changes: bool = False
    previous_agent_count: int = 0
    added_agents_count: int = 0
    removed_agents_count: int = 0
    change_reason: str | None = None


class StoreCoverageResponse(StrictApiModel):
    active_stores_count: int
    uncovered_stores_count: int
    closed_stores_count: int
    modified_stores_count: int = 0
    items: list[StoreCoverageItem]


class AgentMovementPoint(StrictApiModel):
    month: MonthStr
    active: int
    new: int
    reactivated: int
    churned: int
    net_growth: int = 0
    is_baseline: bool = False


class AgentMovementResponse(StrictApiModel):
    history: list[AgentMovementPoint]


class AgentListItem(StrictApiModel):
    agent: str
    store_name: str | None = None
    firma: str | None = None
    active_in_month: bool
    is_new: bool
    is_reactivated: bool
    total_sales: Decimal
    total_quantity: int
    current_status: AgentCurrentStatus


class AgentListResponse(StrictApiModel):
    items: list[AgentListItem]


class AgentProfileResponse(StrictApiModel):
    agent: str
    first_seen_month: MonthStr
    last_seen_month: MonthStr
    active_months_count: int
    distinct_store_count: int
    distinct_firma_count: int
    distinct_regional_count: int
    distinct_asm_count: int
    months_since_last_seen: int
    reactivation_count: int
    longest_active_streak: int
    career_total_sales: Decimal
    career_total_quantity: int
    avg_monthly_sales: Decimal
    best_month: MonthStr | None
    best_month_sales: Decimal
    current_status: AgentCurrentStatus


class AgentHistoryPoint(StrictApiModel):
    month: MonthStr
    total_sales: Decimal
    total_quantity: int
    receipt_count: int
    active_store_count: int
    is_active: bool


class AgentHistoryResponse(StrictApiModel):
    history: list[AgentHistoryPoint]


class AgentEvaluationOption(StrictApiModel):
    value: str
    label: str


class AgentEvaluationRow(StrictApiModel):
    month: AgentEvaluationPeriod
    firma: str
    site_code: str
    locatie: str
    regional: str
    asm: str
    agent: str
    total_sales: Decimal
    total_quantity: int
    working_days: int
    store_target: Decimal
    store_working_days: int
    target_value: Decimal
    target_pct: Decimal | None
    daily_average: Decimal | None
    peer_daily_average: Decimal | None
    value_reper: Decimal | None
    receipt_count: int
    receipt_2plus_count: int
    bonuri_pct: Decimal | None
    focus_quantity: int
    focus_pct: Decimal | None
    glass_qty: int
    premium_glass_qty: int
    premium_glass_pct: Decimal | None
    target_points: int
    daily_points: int
    value_reper_points: int
    bonuri_points: int
    focus_points: int
    premium_glass_points: int
    total_points: int
    has_red_segment: bool
    qualifier: AgentQualifier


class AgentEvaluationResponse(StrictApiModel):
    months: list[AgentEvaluationOption]
    firmas: list[AgentEvaluationOption]
    asms: list[AgentEvaluationOption]
    stores: list[AgentEvaluationOption]
    rows: list[AgentEvaluationRow]


class AgentEvaluationV2Component(StrictApiModel):
    value: Decimal | None
    reference: Decimal | None = None
    score: Decimal | None
    weight: int
    label: str | None = None


class AgentEvaluationV2Row(StrictApiModel):
    month: AgentEvaluationPeriod
    firma: str
    site_code: str
    locatie: str
    regional: str
    asm: str
    agent: str
    total_sales: Decimal
    forecast_sales: Decimal
    total_quantity: int
    working_days: int
    receipt_count: int
    target_value: Decimal
    target_source: TargetSource
    target_pct: Decimal | None
    target_forecast_pct: Decimal | None
    is_partial: bool
    period_month_count: int
    partial_month_count: int
    final_month_count: int
    forecast_factor: Decimal
    daily_average: Decimal | None
    daily_reference: Decimal | None
    daily_reference_type: DailyReferenceType
    daily_vs_reference_pct: Decimal | None
    value_reper: Decimal | None
    receipt_2plus_count: int
    bonuri_pct: Decimal | None
    focus_quantity: int
    focus_pct: Decimal | None
    glass_qty: int
    premium_glass_qty: int
    premium_glass_pct: Decimal | None
    trend_daily_pct: Decimal | None
    trend_direction: TrendDirection
    eligibility_status: EligibilityStatus
    confidence_flags: list[str]
    target_score: Decimal | None
    daily_score: Decimal | None
    bonuri_score: Decimal | None
    focus_score: Decimal | None
    premium_glass_score: Decimal | None
    value_reper_score: Decimal | None
    total_score: Decimal | None
    max_score: int = 100
    rating: AgentRating


class AgentEvaluationV2Response(StrictApiModel):
    months: list[AgentEvaluationOption]
    firmas: list[AgentEvaluationOption]
    asms: list[AgentEvaluationOption]
    stores: list[AgentEvaluationOption]
    rows: list[AgentEvaluationV2Row]
